templates/internal/billing/billing.go

```diff
@@ -0,0 +1,175 @@
+	for stripePriceIterator.Next() {
+		stripePrice := stripePriceIterator.Current().(*Stripe.Price)
+		amount := float64(stripePrice.UnitAmount) / 100
+		displayPrice := fmt.Sprintf("%s $%.2f/%s", strings.ToUpper(string(stripePrice.Currency)), amount, string(stripePrice.Recurring.Interval))
+		sub := Subscriptions{
+			Nickname: stripePrice.Nickname,
+			Interval: string(stripePrice.Recurring.Interval),
+			Type:     string(stripePrice.Type),
+			ID:       stripePrice.ID,
+			Price:    displayPrice,
+		}
+		subs = append(subs, sub)
+	}
+	output, err := json.Marshal(&subs)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	w.Write(output)
+}
+
+func checkout(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "Incorrect method", http.StatusBadRequest)
+		return
+	}
+
+	price := struct {
+		ID string `json:"price_id"`
+	}{}
+	err := json.NewDecoder(r.Body).Decode(&price)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	paymentMethod := "card"
+	paymentMode := "subscription"
+	checkoutQuantity := int64(1)
+	lineItem := &Stripe.CheckoutSessionLineItemParams{
+		Price:    &price.ID,
+		Quantity: &checkoutQuantity,
+	}
+	successURL := backendURL + "/billing/success?session_id={CHECKOUT_SESSION_ID}"
+	cancelURL := backendURL + "/billing/cancel?session_id={CHECKOUT_SESSION_ID}"
+<% if eq (index .Params `userAuth`) "yes" %>
+	authErr, userInfo := auth.GetUserInfoFromHeaders(r)
+	clientReferenceID := userInfo.ID
+	if authErr != nil {
+		http.Error(w, authErr.Error(), http.StatusUnauthorized)
+		return
+	}
+	<%- else %>
+	clientReferenceID := "internal-app-reference-id"
+	<%- end %>
+
+	checkoutParams := &Stripe.CheckoutSessionParams{
+		Mode:               &paymentMode,
+		PaymentMethodTypes: []*string{&paymentMethod},
+		ClientReferenceID:  &clientReferenceID,
+		LineItems:          []*Stripe.CheckoutSessionLineItemParams{lineItem},
+		SuccessURL:         &successURL,
+		CancelURL:          &cancelURL,
+	}
+	session, err := stripe.CheckoutSessions.New(checkoutParams)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	fmt.Fprintf(w, `{"sessionId": "%s"}`, session.ID)
+}
+
+func success(w http.ResponseWriter, r *http.Request) {
+	sessionId := string(r.URL.Query().Get("session_id"))
+
+	session, err := stripe.CheckoutSessions.Get(sessionId, &Stripe.CheckoutSessionParams{})
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	data := map[string]string{
+		"payment_status": string(session.PaymentStatus),
+		"amount":         fmt.Sprintf("%d", session.AmountSubtotal),
+		"currency":       string(session.Currency),
+		"customer":       string(session.CustomerDetails.Email),
+		"reference":      string(session.ClientReferenceID),
+	}
+
+	redirectURL := fmt.Sprintf("%s%s?%s", frontendURL, "/billing/confirmation", mapToQueryString(data))
```

Comment:
Probably better to use the `net/url` library here.. You can call use `Values.Add()` to build up query parameters, then call `Values.Encode()` to get the query string which you can use in a URL object: https://golang.org/pkg/net/url/#Values.Encode

Then you can add the host and path and use `URL.String()` to get the string for it.